106-auto-rebalance-on-imbalanced-clusters.md

```diff
@@ -0,0 +1,471 @@
+3. In case, a rebalance is already ongoing and another configmap related to goal violation is detected, then the operator will just ignore the configmap and delete it.
+
+The `KafkaAutoRebalancingReconciler.reconcile()` loads the `Kafka.status.autoRebalance` content:
+
+* `state`: is the FSM state.
+* `lastTransitionTime`: when the transition to that state happened.
+* `modes`: sets the mode as `skew`
+
+The FSM is initialized based on the `state` field.
+
+Let's see what happens during the auto-rebalancing process when the FSM starts from the **Idle** state and transitions to **RebalanceOnAnomalyDetection**
+
+#### Idle
+
+This state is set since the beginning when a `Kafka` custom resource is created with the `spec.cruiseControl.autoRebalance` field.
+It is also the end state of a previous successfully completed or failed auto-rebalancing.
+In case of successful completion, once the rebalance moves to `Ready` state, we will delete the KafkaRebalance and the configmap associated with the rebalance by matching the `anomalyId` suffix in their names and then update the `auto-rebalance` state to `Idle`.
+In case of failed auto-rebalancing, once the rebalance moves to `NotReady` state, we will foloow the same procedure we used in successful completion.
+In this state, the operator removes the finalizer and deletes the corresponding "actual" `KafkaRebalance` custom resource.
+
+#### RebalanceOnAnomalyDetection
+
+In this state, an anomaly was detected and a corresponding configmap was generated by the notifier.
+
+ A KafkaRebalance resource will now be applied to the cluster to fix the imbalanced cluster. This kafka rebalance will be based on the template provided by the user, if no template is provided then the kafkaRebalance will be created with default configurations.
+
+```mermaid
+flowchart TB
+  A[KafkaRebalanceState] --> B{if state}
+  B --> C[Ready]
+  C -- queued scale Down --> D[RebalanceOnScaleDown]
+  C -- queued scale up --> E[RebalanceOnScaleUp]
+  C -- queued rebalance --> F[RebalanceOnAnomalyDetection]
+  B --> G[PendingProposal, ProposalReady or Rebalancing]
+  G --> H{{rebalancing is running}}
+  B --> I[NotReady]
+  I --> J{{Transition to Idle state and delete rebalance resource and configmap}}
+```
+
+Checking the current `KafkaRebalance` status:
+
+* if `Ready`, the rebalance was successful.
+  * if there is a queued rebalancing scale down (`Kafka.status.autoRebalance.modes[remove-brokers]` exists), start the rebalancing scale down and transition to **RebalanceOnScaleDown**.
+  * if there is a queued rebalancing scale up (`Kafka.status.autoRebalance.modes[add-brokers]` exists), start the rebalancing scale up and transition to **RebalanceOnScaleUp**.
+  * If no queued rebalancing scale down or scale up, just transition to **Idle**, clean `Kafka.status.autoRebalance.modes`, delete the "actual" `KafkaRebalance` custom resource and also the configmap that triggered the rebalance by matching the `anomalyId` suffix in their names.
```

Comment:
I think it is still important to be able to see when auto-rebalancing was triggered, for debugging purposes. Does Strimzi emit metrics for when it triggers auto-rebalancing and in what mode?